### PR TITLE
[AIR] Make `Checkpoint.get_preprocessor` faster

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Previously, `get_preprocessor` would always serialize the Checkpoint into a dictionary first. This is incredibly wasteful and causes huge memory usage and runtime with large directory-based Checkpoints. This PR changes the logic to first see if a directory Checkpoint should be loaded into a dictionary or not in order to obtain the preprocessor.

Context: I had ran into it when trying to do predictions with 25 GB Hugging Face model. `HuggingFacePredictor` calls `get_preprocessor` internally, and that takes ages to complete and almost caused an OOM for me - and all of that is unnecessary as the preprocessor has to be loaded from a file anyway.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/checkpoint.py [hunk 1]
```diff
@@ -760,12 +760,7 @@ def __fspath__(self):
             "Use `Checkpoint.to_directory()` or `Checkpoint.as_directory()` instead."
         )
 
-    def get_preprocessor(self) -> Optional["Preprocessor"]:
-        """Return the saved preprocessor, if one exists."""
-
-        # The preprocessor will either be stored in an in-memory dict or
-        # written to storage. In either case, it will use the PREPROCESSOR_KEY key.
-
+    def _get_preprocessor(self) -> Optional["Preprocessor"]:
```
Comment (krfricke) [inline python/ray/air/checkpoint.py:763]: Do we need this internal function? We don't use it anywhere else?

File: python/ray/air/checkpoint.py [hunk 2]
```diff
@@ -777,6 +772,32 @@ def get_preprocessor(self) -> Optional["Preprocessor"]:
...
+                if _is_persisted_directory_checkpoint(checkpoint_path):
+                    # If this is a persisted directory checkpoint, then we load the
+                    # files from the temp directory created by the context.
+                    # That way we avoid having to download the files twice.
+                    loaded_checkpoint = self.from_directory(checkpoint_path)
+                    preprocessor = loaded_checkpoint._get_preprocessor()
+                else:
+                    preprocessor = load_preprocessor_from_dir(checkpoint_path)
+        else:
+            preprocessor = self._get_preprocessor()
```
Comment (krfricke) [inline python/ray/air/checkpoint.py:797]:
```suggestion
                checkpoint_dict = self.to_dict()
                preprocessor = checkpoint_dict.get(PREPROCESSOR_KEY, None)
```

Shouldn't this be enough?

**Review by krfricke — CHANGES_REQUESTED** [review]
Minor questions

File: python/ray/air/checkpoint.py (hunk 1 above)
Comment (Yard1) [reply python/ray/air/checkpoint.py:763]: It's used below

File: python/ray/air/checkpoint.py (hunk 2 above)
Comment (Yard1) [reply python/ray/air/checkpoint.py:797]: We may have a bytes checkpoint created from a directory, in which case this would fail.

**Review by krfricke — APPROVED** [review]
Thanks!

**Pull request merged**